FFmpegInterop/NALUSampleProvider.cpp

```diff
@@ -0,0 +1,255 @@
+		videoProp.Insert(MF_NALU_LENGTH_SET, PropertyValue::CreateUInt32(static_cast<uint32_t>(true)));
+
+		if (!m_spsPpsData.empty())
+		{
+			videoProp.Insert(MF_MT_MPEG_SEQUENCE_HEADER, PropertyValue::CreateUInt8Array({ m_spsPpsData.data(), m_spsPpsData.data() + m_spsPpsData.size() }));
+		}
+	}
+
+	tuple<IBuffer, int64_t, int64_t, map<GUID, IInspectable>> NALUSampleProvider::GetSampleData()
+	{
+		// Get the next sample
+		AVPacket_ptr packet{ GetPacket() };
+
+		const int64_t pts{ packet->pts };
+		const int64_t dur{ packet->duration };
+		const bool isKeyFrame{ (packet->flags & AV_PKT_FLAG_KEY) != 0 };
+
+		// Transform the sample into the format expected by the decoder
+		auto [buf, naluLengths] = TransformSample(move(packet), isKeyFrame);
+
+		// Set sample properties
+		map<GUID, IInspectable> properties;
+
+		if (isKeyFrame)
+		{
+			properties[MFSampleExtension_CleanPoint] = PropertyValue::CreateUInt32(true);
+		}
+
+		const uint8_t* naluLengthsBuf{ reinterpret_cast<const uint8_t*>(naluLengths.data()) };
+		const size_t naluLengthsBufSize{ sizeof(decltype(naluLengths)::value_type) * naluLengths.size() };
+		properties[MF_NALU_LENGTH_INFORMATION] = PropertyValue::CreateUInt8Array({ naluLengthsBuf, naluLengthsBuf + naluLengthsBufSize });
+
+		return { move(buf), pts, dur, move(properties) };
+	}
+
+	tuple<IBuffer, vector<uint32_t>> NALUSampleProvider::TransformSample(_Inout_ AVPacket_ptr packet, _In_ bool isKeyFrame)
+	{
+		bool copySpsPpsData{ isKeyFrame && !m_spsPpsData.empty() };
+		const bool writeToBuf{ copySpsPpsData || (!m_isBitstreamAnnexB && m_naluLengthSize != sizeof(NALU_START_CODE)) };
+
+		vector<uint8_t> buf;
+		if (writeToBuf)
+		{
+			size_t sampleSize{ static_cast<uint32_t>(packet->size) };
+
+			if (isKeyFrame)
+			{
+				// Reserve space for SPS/PPS data
+				sampleSize += m_spsPpsData.size();
+			}
+
+			if (!m_isBitstreamAnnexB && m_naluLengthSize > sizeof(NALU_START_CODE))
+			{
+				// Reserve additional space since NALU start code is longer than NALU length size
+				sampleSize += MAX_NALU_NUM_SUPPORTED * (sizeof(NALU_START_CODE) - m_naluLengthSize);
+			}
+
+			buf.reserve(sampleSize);
+		}
+
+		vector<uint32_t> naluLengths;
+		naluLengths.reserve(MAX_NALU_NUM_SUPPORTED);
+
+		const uint8_t naluPrefixLength{ m_isBitstreamAnnexB ? sizeof(NALU_START_CODE) : m_naluLengthSize };
+		for (uint32_t i{ 0 }; i + naluPrefixLength <= static_cast<uint32_t>(packet->size);)
+		{
+			// Get the length of the next NALU
+			uint32_t naluLength{ 0 };
+			if (m_isBitstreamAnnexB)
+			{
+				naluLength = GetAnnexBNaluLength(packet->data + i, packet->size - i);
+			}
+			else
+			{
+				naluLength = GetAVCNaluLength(packet->data + i, packet->size - i, m_naluLengthSize);
+			}
+
+			if (copySpsPpsData)
+			{
+				// Check if this NALU is an AUD (access unit delimiter). The AUD must be the first NALU in the sample.
+				if (naluLength < naluPrefixLength || packet->data[naluPrefixLength] != NALU_TYPE_AUD)
```

Comment:
>naluLength < naluPrefixLength [](start = 8, length = 29)

Shouldn't this be "naluLength == 0"? The length of an AUD NALU can be less than the prefix length - the payload just contains the NALU header (1 byte) + primary_pic_type (3 bits) + padding for byte alignment (5 bits).